domain/schema/model.go

```diff
@@ -558,11 +599,92 @@ INSERT INTO storage_kind VALUES
     (0, 'block', 'Allows for the creation of raw storage volumes'), 
     (1, 'filesystem', 'Provides a hierarchical file storage system');
 
+-- This table stores storage constraint values for each named storage item
+-- defined by the application's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+CREATE TABLE application_storage_constraint (
+    application_uuid TEXT NOT NULL,
+    charm_uuid       TEXT NOT NULL,
+    storage_name     TEXT NOT NULL,
+    -- These attributes are filled in by sourcing data from:
+    -- user supplied, model config, charm config, opinionated fallbacks.
+    -- By the time the row is written, all values are known.
+    -- Constraint value types (pool, size, count) hitherto have
+    -- been fixed (since first implemented). We don't envisage
+    -- any change to how these are modelled.
+    --
+    -- Note: one might wonder why storage_pool below is not a
+    -- FK to a row defined in the storage pool table. This value
+    -- can also be one of the pool types. As with the comment on the
+    -- type column in the storage pool table, it's problematic to use a lookup
+    -- with an ID. Storage pools, once created, cannot be renamed so
+    -- this will not be able to become "orphaned".
+    storage_pool TEXT NOT NULL,
+    size         INT  NOT NULL,
+    count        INT  NOT NULL,
+    CONSTRAINT       fk_application_storage_constraint_application
+        FOREIGN KEY  (application_uuid)
+        REFERENCES   application(uuid),
+    CONSTRAINT       fk_application_storage_constraint_charm_storage
+        FOREIGN KEY  (charm_uuid, storage_name)
+        REFERENCES   charm_storage(charm_uuid, name),
+    PRIMARY KEY (application_uuid, charm_uuid, storage_name)
+);
+
+-- Note that this is not unique; it speeds access by application.
+CREATE INDEX idx_application_storage_constraint
+ON application_storage_constraint (application_uuid);
+
+-- This table stores storage constraint values for each named storage item
+-- defined by the unit's current charm. If the charm is updated, then
+-- so too will be the rows in this table to reflect the current charm's
+-- storage definitions.
+-- Note: usually we just get the storage constraints off the application
+-- but need to allow for a unit's charm to temporarily diverge from that
+-- of its application.
+CREATE TABLE unit_storage_constraint (
+    unit_uuid    TEXT NOT NULL,
+    charm_uuid   TEXT NOT NULL,
```

Comment:
A unit will have some link to its current charm already, most likely in the actual unit table. There's no need to store it here; it makes for confusing uniqueness possibilities.

Reply:
This was done because you requested a FK to the charm storage row to which this constraint pertains. The FK is the (charm_uuid, storage name) tuple. Charm storage rows do not have a uuid - they are a child of a parent charm and are identified by their parent charm uuid and name.
The PR originally just had storage name and omitted the charm uuid.

Reply:
I get this now. 

I would be inclined to have a single primary key on `charm_storage` (just an auto-incrementing ID if you don't want to model it as an outright entity), then the FK for `charm_storage_property`, `application_storage_constraint` and `unit_storage_constraint` is a single instead of composite key.

Reply:
This would then model this 1..N differently than we do for other things where the N side has a composite primary key of the parent record and the unique name. If we are happy with that divergence in how we model these things, I can do that an auto inc ID but really prefer consistency. Yes, the approach here does in effect dup the charm_uuid, but only transitively. IMO it's an acceptable trade off for modelling consistency.
It's definitely not an outright entity; it's the same as for example a storage pool attribute where the composite PK is pool_uuid and attr name; storage pool attribute is not an entity in its own right, it doesn't get a uuid; it's part of a storage pool and doesn't exist on its own.